src/str.rs

```diff
@@ -433,186 +445,229 @@ pub(crate) fn parse_str_radix_n(str: &str, radix: u32) -> Result<Decimal, crate:
+                    has_digit = true;
+
+                    data64 = data64 * RADIX as u64 + digit as u64;
+                    scale += passed_decimal_point as u8;
+
+                    if scale >= estimated_max_precision(RADIX) {
+                        if !bytes.is_empty() {
+                            return maybe_round_radix::<RADIX>(
+                                data64 as u128,
+                                bytes[0],
+                                scale,
+                                passed_decimal_point,
+                                negative,
+                            );
+                        }
+                        break;
+                    }
                 }
-            }
-            b'.' => {
-                if digits_before_dot >= 0 {
-                    return Err(Error::from("Invalid decimal: two decimal points"));
+                b'.' => {
+                    if passed_decimal_point {
+                        return Err(Error::from("Invalid decimal: two decimal points"));
+                    }
+                    passed_decimal_point = true;
                 }
-                digits_before_dot = coeff.len() as i32;
-                offset += 1;
-                len -= 1;
-            }
-            b'_' => {
-                // Must start with a number...
-                if coeff.is_empty() {
-                    return Err(Error::from("Invalid decimal: must start lead with a number"));
+                b'_' => {
+                    // Must start with a number...
+                    if !has_digit {
+                        return Err(Error::from("Invalid decimal: must start lead with a number"));
+                    }
                 }
-                offset += 1;
-                len -= 1;
+                _ => return Err(Error::from("Invalid decimal: unknown character")),
             }
-            _ => return Err(Error::from("Invalid decimal: unknown character")),
+            bytes = &bytes[1..];
+        } else {
+            break;
         }
     }
 
-    // If we exited before the end of the string then do some rounding if necessary
-    if maybe_round && offset < bytes.len() {
-        let next_byte = bytes[offset];
-        let digit = match next_byte {
-            b'0'..=b'9' => {
-                if next_byte > max_n {
-                    return Err(Error::from("Invalid decimal: invalid character"));
-                }
-                u32::from(next_byte - b'0')
-            }
-            b'a'..=b'z' => {
-                if next_byte > max_alpha_lower {
+    let data: u128 = data64 as u128;
+
+    if !bytes.is_empty() {
+        return handle_full_128_radix::<RADIX>(data, bytes, scale, passed_decimal_point, negative);
+    } else if !has_digit {
+        return Err(Error::from("Invalid decimal: no digits found"));
+    }
+
+    handle_data_radix(data, scale, negative)
+}
+
+const fn midpoint_for_radix(radix: u8) -> u8 {
+    if radix & 0x1 == 1 {
```

Comment:
is there an is_even function?

Reply:
there isn't, but yes we can add

Reply:
ah then no worries just figured something builtin could use that